lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -606,70 +688,127 @@ LogicalResult ComponentOp::verify() {
+WiresOp calyx::CombComponentOp::getWiresOp() {
+  auto *body = getBodyBlock();
+  auto opIt = body->getOps<WiresOp>().begin();
+  return *opIt;
+}
 
-  // Record the port names and number of input ports of the component.
-  result.addAttribute("portNames", builder.getArrayAttr(portNames));
-  result.addAttribute("portDirections",
-                      direction::packAttribute(builder.getContext(),
-                                               portIOTypes.first.size(),
-                                               portIOTypes.second.size()));
-  // Record the attributes of the ports.
-  result.addAttribute("portAttributes", builder.getArrayAttr(portAttributes));
+/// A helper function to return a filtered subset of a comb component's ports.
+template <typename Pred>
+static SmallVector<PortInfo> getFilteredPorts(CombComponentOp op, Pred p) {
+  SmallVector<PortInfo> ports = op.getPortInfo();
+  llvm::erase_if(ports, p);
+  return ports;
+}
 
-  // Create a single-blocked region.
-  Region *region = result.addRegion();
-  Block *body = new Block();
-  region->push_back(body);
+SmallVector<PortInfo> CombComponentOp::getInputPortInfo() {
+  return getFilteredPorts(
+      *this, [](const PortInfo &port) { return port.direction == Output; });
+}
 
-  // Add all ports to the body.
-  body->addArguments(portTypes, SmallVector<Location, 4>(
-                                    portTypes.size(), builder.getUnknownLoc()));
+SmallVector<PortInfo> CombComponentOp::getOutputPortInfo() {
+  return getFilteredPorts(
+      *this, [](const PortInfo &port) { return port.direction == Input; });
+}
 
-  // Insert the WiresOp and ControlOp.
-  IRRewriter::InsertionGuard guard(builder);
-  builder.setInsertionPointToStart(body);
-  builder.create<WiresOp>(result.location);
-  builder.create<ControlOp>(result.location);
+void CombComponentOp::print(OpAsmPrinter &p) {
+  printComponentInterface(p, *this);
 }
 
-void ComponentOp::getAsmBlockArgumentNames(
+ParseResult CombComponentOp::parse(OpAsmParser &parser,
+                                   OperationState &result) {
+  return parseComponentInterface(parser, result);
+}
+
+LogicalResult CombComponentOp::verify() {
+  // Verify there is exactly one wires operation.
+  auto wIt = getBodyBlock()->getOps<WiresOp>();
+  if (std::distance(wIt.begin(), wIt.end()) != 1)
+    return emitOpError() << "requires exactly one "
+                         << WiresOp::getOperationName() << " op.";
+
+  // Verify there is not a control operation.
+  auto cIt = getBodyBlock()->getOps<ControlOp>();
+  if (std::distance(cIt.begin(), cIt.end()) != 0)
+    return emitOpError() << "requires exactly one `"
+                         << ControlOp::getOperationName() << "` op.";
+
+  // Verify the component actually does something: has continuous assignments.
+  bool hasNoAssignments =
+      getWiresOp().getBodyBlock()->getOps<AssignOp>().empty();
+  if (hasNoAssignments)
+    return emitOpError(
```

Comment:
nit: missing tests for all of the `emitOpError`s introduced in this commit.

Reply:
Added, thanks for pointing this out.